File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        mp = {}
        for num in nums:
            mp[num] = mp.get(num,0)+1
        
        arr = []
        for s,v in mp.items():
            arr.append([v,s])
        
        arr.sort()

        res = []
        
        while len(res) < k:
            res.append(arr.pop()[1])
      
        return res


        

File: Data Structures & Algorithms/task-scheduling/submission-2.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        vector<int> arr(26,0);
        for(char ch : tasks){
            arr[ch-'A']++;
        }
        priority_queue<int,vector<int>> maxHeap;
        for(int cnt : arr){
            if(cnt > 0){
                maxHeap.push(cnt);
            }
        }
        int t = 0;
        queue<pair<int,int>> q;
        while(!maxHeap.empty() || !q.empty()){
            t++;
            if(maxHeap.empty()){
                t = q.front().second;
            }else{
                int num = maxHeap.top()-1;
                maxHeap.pop();
                if(num > 0){
                    q.push({num,t+n});
                }
            }
            if(!q.empty() && t == q.front().second){
                int k = q.front().first;
                q.pop();
                maxHeap.push(k);
            }
        }
        return t;
    }
};
